### PR TITLE
ddl: fix changing index prefixed column type to non-prefixed

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/25002, https://github.com/pingcap/tidb/issues/24572 

Problem Summary: 
```sql
CREATE TABLE `t` (
  `a` text DEFAULT NULL,
  UNIQUE KEY `a` (`a`(5))
);
alter table t change column a a int;
```

The index prefix length of `a` should not be `5` since the type `int` cannot be prefixed.

```sql
create table t(a char(250), unique key(a(10)));
alter table t modify a char(9);
```

The new index prefix length of `a` less than the original `10`, the new index should not be prefixed. 

In the above 2 cases, the index column length is not updated according to the new column type.

### What is changed and how it works?

What's Changed:

- Set the length of the target column in the new index to `types.UnspecifiedLength`.
- Update the index check part.
- Add a simple test.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

NA

### Release note 

- Fix an issue that cannot change index prefixed column type to non-prefixed one.

### CODE REVIEW

File: ddl/column_type_change_test.go [hunk 1]
```diff
@@ -1818,3 +1819,37 @@ func (s *testColumnTypeChangeSuite) TestChangeIntToBitWillPanicInBackfillIndexes
...
+	showCreateTable = tk.MustQuery("show create table t").Rows()[0][1].(string)
+	c.Assert(strings.Contains(showCreateTable, "UNIQUE KEY `idx` (`a`(2),`b`)"), IsTrue,
+		Commentf("%s", showCreateTable))
+
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t(a char(250), unique key idx(a(10)));")
+	tk.MustExec("alter table t modify a char(9);")
+	showCreateTable = tk.MustQuery("show create table t").Rows()[0][1].(string)
+	c.Assert(strings.Contains(showCreateTable, "UNIQUE KEY `idx` (`a`)"), IsTrue,
+		Commentf("%s", showCreateTable))
```
Comment (wjhuang2016) [inline ddl/column_type_change_test.go:1854]:
``` sql
create table t(a varchar(700), key(a(700)));
alter table t change column a a tinytext;
```
Please add this case.

Comment (tangenta) [reply ddl/column_type_change_test.go:1854]: Let's wait for #24959.

Comment (tangenta) [reply ddl/column_type_change_test.go:1919]: Updated. The behavior is consistent with MySQL 8.0.

**Review by zimulala — APPROVED** [review]
LGTM

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**